src/client/pydaos/torch/torch_shim.c

```diff
@@ -0,0 +1,887 @@
+	ssize_t            rc     = 0;
+	struct dfs_handle *hdl    = NULL;
+	char              *path   = NULL;
+	dfs_obj_t         *obj    = NULL;
+	PyObject          *buffer = NULL;
+
+	RETURN_NULL_IF_FAILED_TO_PARSE(args, "LsO", &hdl, &path, &buffer);
+	assert(hdl->dfs != NULL);
+
+	if (!PyObject_CheckBuffer(buffer)) {
+		PyErr_SetString(PyExc_TypeError,
+				"Expected an object that supports the buffer protocol");
+		return NULL;
+	}
+
+	Py_buffer bview;
+	if (PyObject_GetBuffer(buffer, &bview, PyBUF_WRITE) == -1) {
+		PyErr_SetString(PyExc_BufferError, "Buffer is not writable");
+		return NULL;
+	}
+
+	/*
+	  Since python can use buffer like objects that might not have contiguous memory layout,
+	  let's put a guardrail accepting only buffers with contiguous memory region
+	*/
+	if (!PyBuffer_IsContiguous(&bview, 'C')) {
+		PyErr_SetString(PyExc_BufferError, "Buffer is not contiguous");
+		PyBuffer_Release(&bview);
+		return NULL;
+	}
+
+	daos_size_t read = bview.len;
+
+	d_iov_t     iov;
+	d_iov_set(&iov, bview.buf, read);
+
+	d_sg_list_t sgl = {
+	    .sg_nr     = 1,
+	    .sg_nr_out = 0,
+	    .sg_iovs   = &iov,
+	};
+
+	rc = dfs_lookup(hdl->dfs, path, O_RDONLY, &obj, NULL, NULL);
+	if (rc) {
+		D_ERROR("Could not lookup '%s': %s (rc=%ld)", path, strerror(rc), rc);
+		rc = -rc;
+		goto out;
+	}
+
+	rc = dfs_read(hdl->dfs, obj, &sgl, 0 /* offset */, &read, NULL);
+	if (rc) {
+		rc = -rc;
+		goto out;
+	}
+	rc = read;
+
+out:
+	PyBuffer_Release(&bview);
+
+	if (obj) {
+		int rc2 = dfs_release(obj);
+		if (rc2) {
+			D_ERROR("Could not release object '%s': %s (rc=%d)", path, strerror(rc2),
+				rc2);
+		}
+	}
+
+	return PyLong_FromLong(rc);
+}
+
+/* describes the IO operation */
+struct io_op {
+	daos_event_t ev;
+
+	dfs_obj_t   *obj;
+	PyObject    *item;
+
+	/* Purely for debug purpose: should not be freed as it's not the owner of the data */
+	const char  *path;
+
+	daos_size_t  size;
+	int          err;
+
+	d_iov_t      iov;
+	d_sg_list_t  sgl;
+
+	/* Buffer view that implements Python Buffer Protocol, has to be release after use */
+	Py_buffer    buf_view;
+};
+
+static int
+start_read_op(struct dfs_handle *hdl, PyObject *item, struct io_op *op)
+{
+	assert(op != NULL);
+
+	int           rc  = 0;
+	daos_event_t *evp = &op->ev;
+
+	PyObject     *py_path = PyTuple_GetItem(item, 0);
+	PyObject     *py_buff = PyTuple_GetItem(item, 1);
+
+	if (py_path == NULL || py_buff == NULL) {
+		D_WARN("Each tuple must contain exactly two elements: path and bytearray");
+		return EINVAL;
+	}
+
+	const char *path = PyUnicode_AsUTF8(py_path);
+	if (path == NULL) {
+		D_WARN("First element of a tuple does not look like a path");
+		return EINVAL;
+	}
+
+	if (PyObject_GetBuffer(py_buff, &op->buf_view, PyBUF_WRITE) == -1) {
+		D_WARN("Buffer is not writable");
+		return EINVAL;
+	}
+
+	if (!PyBuffer_IsContiguous(&op->buf_view, 'C')) {
+		D_WARN("Buffer for '%s' is not contiguous", path);
+		rc = EINVAL;
+		goto out;
+	}
+
+	rc = daos_event_init(evp, hdl->eq, NULL);
```

Comment:
can you please explain the threading model for reads?
it's not a good idea performance wise to use a single EQ if you have multiple threads on that same EQ. otherwise you are limited to a single network context and your performance will suffer.

Reply:
By default Dataset is single threaded (more like single process in python), __getitem__() and __getitems__() are regular blocking calls. During the dataset creation the connection is established to DAOS and the even queue is created, that is going to be used to fetch the data.
If multiprocessing is enabled, Dataset provides the worker_init method which worker processes are calling upon their startup, that's where torch_reinit re-uses the connections and create an even queue per these worker processes.
There's no internal multithreading inside the shim module - it's driven by the outside torch.utils.DataLoader.
So in theory if DataLoader is configured to have 8 readers - 8 even queues are going to be created and no performance degradation should occur.

Reply:
ok thanks for the explanation. then this seems fine.